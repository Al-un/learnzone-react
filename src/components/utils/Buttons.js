import React from "react";
import { Link } from "react-router-dom";
import PropTypes from "prop-types";

/**
 * Create button can:
 * - redirect to `new` URL
 * - TODO...
 */
const CreateButton = props => {
  // URL redirection
  if (props.url) {
    return (
      <Link to={props.url} className="btn btn-outline-primary">
        <span className="fas fa-plus" />
        <span>{props.value}</span>
      </Link>
    );
  }
  // TODO: other create buttons
  else {
    return <div />;
  }
};

CreateButton.propTypes = {
  url: PropTypes.string,
  value: PropTypes.string
};

CreateButton.defaultProps = {
  value: "Create"
};

/**
 * Delete button can only delete via an Ajax request
 */
const DeleteButton = props => (
  <button className="btn btn-danger btn-sm" onClick={props.deleteFunc}>
    <span className="fas fa-trash" />
    <span>Delete</span>
  </button>
);

DeleteButton.propTypes = {
  deleteFunc: PropTypes.func.isRequired
};

export { CreateButton, DeleteButton };
