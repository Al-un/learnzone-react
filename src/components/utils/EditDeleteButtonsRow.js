import React from "react";
import { Link } from "react-router-dom";
import auth from "../../services/auth";
import PropTypes from "prop-types";

/**
 * Rendering a line of *edit* and *delete* buttons if the current user is
 *  authenticated
 */
export default class EditDeleteButtonsRow extends React.Component {
  /**
   * Ensure confirmation before deletion. Confirmation message can be customized
   * via the `deleteMsg` props
   */
  handleDeletion = () => {
    let confirmMsg = this.props.deleteMsg || "Confirm deletion?";
    let confirmDel = window.confirm(confirmMsg);
    if (confirmDel) {
      this.props.deleteEntity();
    }
  };

  /**
   * Rendering with a bootstrap flex
   */
  render() {
    return auth.isAuthenticated() ? (
      // allowed to edit/delete
      <div className="d-flex justify-content-between">
        <Link to={this.props.editPath} className="btn btn-sm btn-info">
          <span className="fas fa-edit" />
          Edit
        </Link>

        <button className="btn btn-danger btn-sm" onClick={this.handleDeletion}>
          <span className="fas fa-trash" />
          Delete
        </button>
      </div>
    ) : (
      // TODO: how to display it?
      <div className="unauthorized" />
    );
  }
}

EditDeleteButtonsRow.propTypes = {
  editPath: PropTypes.string.isRequired,
  deleteEntity: PropTypes.func.isRequired,
  deleteMsg: PropTypes.string
};
