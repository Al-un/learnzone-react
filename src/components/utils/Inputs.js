import React from "react";
import PropTypes from "prop-types";

const IdHiddenInput = props => {
  return props.id ? <input type="hidden" name="id" value={props.id} /> : "";
};

IdHiddenInput.propsType = {
  id: PropTypes.number
};

export { IdHiddenInput };
