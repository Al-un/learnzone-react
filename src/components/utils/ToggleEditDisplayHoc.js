import React from "react";

/**
 * This HoC switches between a display only mode and an edition form.
 * @param {Component} FormComponent edition mode
 * @param {Component} DetailComponent display mode
 * @param {Function} submitEntityFn function to call whenever the form is submitted
 */
function toggleEditDisplayHoc(FormComponent, DetailComponent, submitEntityFn) {
  return class extends React.Component {
    constructor(props) {
      super(props);
      console.log(`props: ${JSON.stringify(props)}`);
      this.state = {
        editing: this.isEditing(props.crud),
        entity: props.entity
      };
      // console.log(`loading entity ${JSON.stringify(this.state)}`);
    }

    /**
     *
     * @param {String} crud crud status
     * @return true if crud triggers edition
     */
    isEditing(crud) {
      return "new" === crud || "edit" === crud;
    }

    /**
     * Switch to edition mode
     */
    enableEditionMode = event => {
      this.setState({ editing: true });
    };

    /**
     * Switch to display only
     */
    disableEditionMode = event => {
      this.setState({ editing: false });
    };

    /**
     * Submit the form from the FormComponent. The submitEntityFn handles
     * creation or update distinction
     */
    handleFormSubmit = event => {
      event.preventDefault();
      this.disableEditionMode();
      console.log(`submitting entity ${JSON.stringify(this.state)}`);
      submitEntityFn(this.state.entity);
    };

    /**
     * Update the local state whenever a input is modified
     */
    handleValueChange = event => {
      // console.log(`New value of ${event.target.name} is ${event.target.value}`);
      this.setState({
        entity: {
          ...this.state.entity,
          [event.target.name]: event.target.value
        }
      });
      // console.log(`updating entity ${JSON.stringify(this.state)}`);
    };

    /**
     * Conditionally render the form or the display only
     */
    render() {
      return this.state.editing ? (
        <FormComponent
          entity={this.state.entity}
          handleFormSubmit={this.handleFormSubmit}
          handleValueChange={this.handleValueChange}
        />
      ) : (
        <DetailComponent
          entity={this.state.entity}
          enableEditionMode={this.enableEditionMode}
        />
      );
    }
  };
}

export default toggleEditDisplayHoc;
