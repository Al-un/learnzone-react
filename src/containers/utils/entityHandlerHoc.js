import React from "react";
import CRUD from "../../api/crud";

/**
 *
 * @param {*} WrappedComponent
 * @param {Object} entityFunctions Must define "new" and "load(id)" functions
 */
function entityHandlerHoc(WrappedComponent, entityFunctions) {
  return class extends React.Component {
    constructor(props) {
      super(props);

      this.state = {
        entity: this.props.crud === CRUD.NEW ? entityFunctions.new() : undefined
      };
      console.log(
        `[entityHandlerHoc.construct] for ${
          WrappedComponent.name
        } with props: ${JSON.stringify(props)}`
      );
    }

    componentDidMount() {
      let id = this.props.match.params.id;
      if (id) {
        console.log(`${WrappedComponent.name} loading #${id}`);
        entityFunctions
          .load(this.props.match.params.id)
          .then(response => response.json())
          .then(data => this.setState({ entity: data }));
      } else {
        console.log("No loading entity");
      }
    }

    render() {
      return this.state.entity ? (
        <div>
          <WrappedComponent
            entity={this.state.entity}
            editing={this.state.editing}
            crud={this.props.crud}
          />
        </div>
      ) : (
        <div>Loading...</div>
      );
    }
  };
}

export default entityHandlerHoc;
