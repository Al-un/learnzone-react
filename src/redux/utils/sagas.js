/**
 * Extract the post processing function from the meta parameter only if post
 * processing is a function
 * @param {*} action
 */
export function extractPostProcessing(action) {
  if (!action.meta) {
    return null;
  }
  // Post processing
  const postProcessing = action.meta.postProcessing;
  return postProcessing instanceof Function ? postProcessing : null;
}
