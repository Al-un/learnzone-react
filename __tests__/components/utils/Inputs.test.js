import React from "react";
import renderer from "react-test-renderer";

import { IdHiddenInput } from "../../../src/components/utils/Inputs";

/** ----------------------------------------------------------------------------
 * IdHiddenInput
 */
const invalidEntity = { name: "name" };
const validEntity = { id: "4", name: "name" };
describe("ID Hidden input", () => {
  describe("With an entity.id", () => {
    it("renders the hidden input", () => {
      const tree = renderer
        .create(<IdHiddenInput id={validEntity.id} />)
        .toJSON();
      expect(tree).toMatchSnapshot();
    });
  });

  describe("Without an entity.id", () => {
    it("renders an empty string", () => {
      const tree = renderer
        .create(<IdHiddenInput id={invalidEntity.id} />)
        .toJSON();
      expect(tree).toMatchSnapshot();
    });
  });
});
