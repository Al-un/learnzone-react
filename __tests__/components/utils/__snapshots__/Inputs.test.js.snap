// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ID Hidden input With an entity.id renders the hidden input 1`] = `
<input
  name="id"
  type="hidden"
  value="4"
/>
`;

exports[`ID Hidden input Without an entity.id renders an empty string 1`] = `""`;
